### PR TITLE
ENT-4286: Simplify exception reporting with system_exit()

### PR DESCRIPTION
**Pull request by DuckBoss**
Card ID: ENT-4286
- Updated system_exit() function to directly accept an exception to
format the exception message using ExceptionMapper before printing it.
- Updated various exception handlers to directly pass exceptions
to system_exit() where ExceptionMapper was used to
format the message of that exception.
- Removed list of exceptions and messages support
and removed related test cases for system_exit() function.
- Added new unit tests for system_exit() function.

### CODE REVIEW

File: test/test_managercli.py [hunk 1]
```diff
@@ -337,6 +337,24 @@ class NoStrException(Exception):
...
+
+    def test_only_exception_unicode(self):
+        ex = ValueError("\u2620 \u2603 \u203D")
+        with Capture() as cap:
+            try:
+                system_exit(1, ex)
+            except SystemExit:
+                pass
+        self.assertEqual("%s\n" % str(ex), cap.err)
+
```
Comment (jirihnidek) [inline test/test_managercli.py:357]: I would like to see here unit test for the case, when second argument of `system_exit` is list of exceptions handled by `ExceptionMapper`.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Overall it LGTM.

It seems that we don't use the case, when `system_exit()` accept list of exceptions, but this case is specified in the type hint. Thus, we should support it properly and add at least one unit test for this case.

File: src/subscription_manager/cli.py [hunk 2]
```diff
@@ -190,12 +192,21 @@ def main(self):
...
+
+    :param code: A unix-style system exit code.
+    :type code: int, required
+    :param msgs: A system exit message, or a list of messages and exceptions, or a single exception.
+        This parameter defaults to None.
+    :type msgs: Union[str, List[Union[str, Exception]], Exception], optional
     """
 
     if msgs:
+        if isinstance(msgs, Exception):
```
[the remark this answers is not in the document]
Comment (DuckBoss) [reply src/subscription_manager/cli.py:207]: As per the original criteria of the jira card, I believe that there are no occurrences of lists of exceptions and it should only handle single exceptions being passed through. I will remove the type hint for lists as Pino suggested to simplify this.

File: src/subscription_manager/cli.py [hunk 3]
```diff
@@ -190,12 +192,19 @@ def main(self):
...
+        This parameter defaults to None.
     """
 
     if msgs:
+        if isinstance(msgs, Exception):
+            exception_mapper: ExceptionMapper = ExceptionMapper()
+            msgs = exception_mapper.get_message(msgs)
         if type(msgs) not in [type([]), type(())]:
             msgs = (msgs,)
         for msg in msgs:
```
Comment (jirihnidek) [inline src/subscription_manager/cli.py:210]: I mean following code:

```python
            if isinstance(msg, Exception):
                msg = "%s" % msg
```

**Comment by jirihnidek** [comment]
> > It seems that we don't use the case, when `system_exit()` accept list of exceptions, but this case is specified in the type hint. Thus, we should support it properly and add at least one unit test for this case.
> 
> When I created ENT-4286 some months ago, I found that there were no such cases around. The usual workflow is: in an exception handler, create the user explanation of that exception using `ExceptionMapper`, and then call `system_exit()` with that text. I didn't see any list of exception objects being handled at the same time, and after checking again now I don't see any currently. I'd simply remove the type hint for a list of things, and keep the code simple for now.

Current code in `system_exit()` supports printing of several exceptions, but `ExceptionMapper` is not used for printing these exceptions (see this comment https://github.com/candlepin/subscription-manager/pull/3152/files#r1005332326). Thus, we have two options: 1) Remove more code (from `cli.py` and `test_managercli.py`) and don't support printing of more than one exception at all. 2) Try to use `ExceptionMapper` for all exception. BTW: there is already several unit tests expecting that second argument of `system_exit()` is list of exceptions/strings and Python IDEs print warning about it, because type hint for `system_exit()` doesn't contain list.

> > ... then it could be IMHO possible to remove following code: https://github.com/candlepin/subscription-manager/pull/3152/files#diff-e6264d0f6eaa02488707c738a34b4c5625b5cf4dd6162447cd26e66e220bbf06L211-R223
> 
> Unfortunately this link does not seem to work :-/ Can you please link to the code?

Oh, I added another comment explaining it: https://github.com/candlepin/subscription-manager/pull/3152/files#r1005332326

**Comment by ptoscano** [comment]
> > > It seems that we don't use the case, when `system_exit()` accept list of exceptions, but this case is specified in the type hint. Thus, we should support it properly and add at least one unit test for this case.
> > 
> > When I created ENT-4286 some months ago, I found that there were no such cases around. The usual workflow is: in an exception handler, create the user explanation of that exception using `ExceptionMapper`, and then call `system_exit()` with that text. I didn't see any list of exception objects being handled at the same time, and after checking again now I don't see any currently. I'd simply remove the type hint for a list of things, and keep the code simple for now.
> 
> Current code in `system_exit()` supports printing of several exceptions, but `ExceptionMapper` is not used for printing these exceptions (see this comment https://github.com/candlepin/subscription-manager/pull/3152/files#r1005332326).

Right. What I see is that nowhere a list of elements (strings or exception objects) is passed to `system_exit()`. While I can potentially see a case for a list of strings, I cannot imagine one for exception objects, as usually one is raisen by Python at a time. Hence...

> Thus, we have two options: 1) Remove more code (from `cli.py` and `test_managercli.py`) and don't support printing of more than one exception at all.

... this is what I suggested when creating ENT-4286: it drops dead code, and makes `system_exit()` simpler.

File: src/subscription_manager/cli.py [hunk 4]
```diff
@@ -190,28 +192,20 @@ def main(self):
...
 
 
-def system_exit(code, msgs=None):
+def system_exit(code: int, msg: Union[str, Exception] = None) -> None:
     """
-    Exit with a code and optional message(s). Saved a few lines of code.
+    Exits the process with an exit code and optional message(s).
+
+    :param code: A unix-style system exit code.
+    :param msgs: A system exit message, a list of messages, or a single exception.
```
Comment (ptoscano) [inline src/subscription_manager/cli.py:200]: "list of messages" should be dropped here

**Review by ptoscano — CHANGES_REQUESTED** [review]
Thanks for the updates, it almost LGTM now. I noted a couple of very small bits, nothing problematic anyway.

For reference: I saw the updated declaration of `system_exit()` that still allows `None` value, and the added type hint for it does not seem to allow it:
```python
def system_exit(code: int, msg: Union[str, Exception] = None) -> None:
```
Funnily enough: mypy 0.982 seems to allow that, which surprised me a lot (as `None` is not in the list of allowed types in the `Union`)! Then I checked around all the usages of `system_exit()`, to see whether we need to keep the parameter optional... and sadly we need, because of one specific usage:
https://github.com/candlepin/subscription-manager/blob/8a5e54e13cab89c2d5ca676d11c2870a2adeff4f/src/subscription_manager/cli_command/remove.py#L68-L75
This code is even partially wrong, as it writes error messages using `print()` which goes to stdout, rather than to stderr... I'll fix this manually after this, no need to worry about it in this PR.

File: src/subscription_manager/cli.py [hunk 5]
```diff
@@ -190,28 +192,19 @@ def main(self):
...
 
 
-def system_exit(code, msgs=None):
+def system_exit(code: int, msg: Union[str, Exception, None] = None) -> None:
     """
-    Exit with a code and optional message(s). Saved a few lines of code.
+    Exits the process with an exit code and optional message(s).
+
+    :param code: A unix-style system exit code.
+    :param msgs: A system exit message, a single exception, or none. This parameter defaults to None.
```
Comment (ptoscano) [inline src/subscription_manager/cli.py:200]: typo, s/msgs/msg/

**Review by ptoscano — CHANGES_REQUESTED** [review]
almost LGTM, except for a small typo which i missed earlier (sorry!)

File: src/subscription_manager/cli.py (hunk 5 above)
Comment (DuckBoss) [reply src/subscription_manager/cli.py]: oops, nice catch! Thanks, should be fixed now.

**Review by ptoscano — APPROVED** [review]
Thanks! LGTM now, and it gets my approval.

I'll leave it to @jirihnidek, as he requested changes in previous versions.

**Review by jirihnidek — APPROVED** [review]
LGTM, Thanks for all updates.

File: src/subscription_manager/cli_command/register.py [hunk 6]
```diff
@@ -204,8 +203,7 @@ def _do_auto_attach(self, consumer):
             # set during service.register() call
             attach.AttachService(self.cp).attach_auto(service_level=None)
         except connection.RestlibException as rest_lib_err:
-            mapped_message: str = ExceptionMapper().get_message(rest_lib_err)
-            print_error(mapped_message)
+            print_error(rest_lib_err)
```
Comment (ptoscano) [inline src/subscription_manager/cli_command/register.py:206]: sorry for the late notice of this: this part is not correct, as `print_error()` takes a string as parameter; would it be possible to fix it?

Comment (DuckBoss) [reply src/subscription_manager/cli_command/register.py:206]: Thanks for letting me know, here's a PR to resolve this issue: https://github.com/candlepin/subscription-manager/pull/3162

**Pull request merged**